http/src/v0/root_files.rs

```diff
@@ -58,3 +68,204 @@ async fn cat_inner<T: IpfsTypes>(ipfs: Ipfs<T>, args: CatArgs) -> Result<impl Re
+    arg: String,
+}
+
+pub fn get<T: IpfsTypes>(
+    ipfs: &Ipfs<T>,
+) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
+    path!("get")
+        .and(with_ipfs(ipfs))
+        .and(query::<GetArgs>())
+        .and_then(get_inner)
+}
+
+async fn get_inner<T: IpfsTypes>(ipfs: Ipfs<T>, args: GetArgs) -> Result<impl Reply, Rejection> {
+    use futures::stream::TryStreamExt;
+
+    let mut path = IpfsPath::try_from(args.arg.as_str()).map_err(StringError::from)?;
+    path.set_follow_dagpb_data(false);
+
+    // FIXME: this is here until we have IpfsPath back at ipfs
+    let (cid, _, _) = walk_path(&ipfs, path).await.map_err(StringError::from)?;
+
+    if cid.codec() != Codec::DagProtobuf {
+        return Err(StringError::from("unknown node type").into());
+    }
+
+    Ok(StreamResponse(Unshared::new(walk(ipfs, cid).into_stream())))
+}
+
+fn walk<Types: IpfsTypes>(
+    ipfs: Ipfs<Types>,
+    root: Cid,
+) -> impl TryStream<Ok = Bytes, Error = GetError> + 'static {
+    let mut cache = None;
+    let mut tar_helper = TarHelper::with_buffer_sizes(16 * 1024);
+
+    let mut root = Some(root);
+    let mut maybe_walker: Option<Walker> = None;
+
+    try_stream! {
+        loop {
+            // this mangling with the root and maybe_walker looks like this mainly because
+            //
+            // a) I could not come up with a simpler solution, as we cannot refactor the big match
+            // to a function given the non-uniform yield points
+            //
+            // b) adding more code here, like with an tri-state enum, goes quickly over the current
+            // #[recursion_limit = "512"], which is required due to how async_stream needs to parse
+            // this function (tt-muncher).
+            //
+            // the next will be the root on first round, and pending_links on the next rounds. root
+            // is read with `as_ref` and later dropped by assigning `None` to it on every
+            // iteration.
+            let next = root.as_ref()
+                .or_else(|| maybe_walker.as_ref().map(|w| w.pending_links().0));
+
+            // we either have a cid reference, or we are done
+            let next = match next {
+                Some(cid_ref) => cid_ref,
+                None => return,
+            };
+
+            let Block { data, .. } = ipfs.get_block(next).await?;
+
+            let res = match maybe_walker {
+                None => {
+                    // the HTTP api uses the final Cid name as the root name in the generated tar
+                    // archive; it will be copied to Walker internally so it can be temporary.
+                    let root_name = next.to_string();
+                    Walker::start(&data, &root_name, &mut cache)?
+                },
+                Some(walker) => walker.continue_walk(&data, &mut cache)?,
+            };
+
+            // make sure only first round uses the `root` cid.
+            root = None;
```

Comment:
I wonder if I am just blind to some trick which would allow us to "push" the root Cid in but not walk it yet. Technically we have the cid, name and the depth (0). I can't see any benefits from having the current API, except that it was an obvious one to write.

Reply:
This might be doable.

Reply:
Done